plugins/hls-change-type-signature-plugin/src/Ide/Plugin/ChangeTypeSignature.hs

```diff
@@ -0,0 +1,247 @@
+    -- can we simplify this logic? Just want to make sure ghcSig is a Just value
+    (RealSrcSpan rss _, Just ghcSig) <- listToMaybe [ (locA ss, sigToText ts)
+                                        | L ss (SigD _ ts@(TypeSig _ idsSig _)) <- decls,
+                                        any ((==) declName . occNameString . occName . unLoc) idsSig
+                                      ]
+    pure (rss, ghcSig)
+
+-- | Pretty Print the Type Signature (to validate GHC Error Message)
+sigToText :: Sig GhcPs -> Maybe Text
+sigToText = \case
+  ts@TypeSig {} -> stripSignature (T.pack $ showSDocUnsafe $ ppr ts)
+  _             -> Nothing
+
+stripSignature :: Text -> Maybe Text
+-- for whatever reason incoming signatures MAY have new lines after "::" or "=>"
+stripSignature sig = case T.filter (/= '\n') sig =~ sigRegex :: (Text, Text, Text, [Text]) of
+                        -- No constraints (Monad m =>)
+                         (_, _, _, [sig'])    -> Just $ T.strip sig'
+                        -- Ignore constraints (Monad m =>)
+                         (_, _, _, [_, sig']) -> Just $ T.strip sig'
+                         _                    -> Nothing
+    where
+        -- we want to test everthing after the constraints (GHC never gives us the constraint in the expected signature)
+        sigRegex = ".* :: (.*=>)?(.*)" :: Text
+
+
+changeSigToCodeAction :: Uri -> ChangeSignature -> Command |? CodeAction
+changeSigToCodeAction uri ChangeSignature{..} = InR CodeAction { _title       = mkChangeSigTitle declName actualType
+                                                               , _kind        = Just CodeActionQuickFix
+                                                               , _diagnostics = Just $ List [diagnostic]
+                                                               , _isPreferred = Nothing
+                                                               , _disabled    = Nothing
+                                                               , _edit        = Just $ mkChangeSigEdit uri declSrcSpan (mkNewSignature declName actualType)
+                                                               , _command     = Nothing
+                                                               , _xdata       = Nothing
+                                                               }
+
+mkChangeSigTitle :: Text -> Text -> Text
+mkChangeSigTitle declName actualType = "change signature for ‘" <> declName <> "’ to: " <> actualType
+
+mkChangeSigEdit :: Uri -> RealSrcSpan -> Text -> WorkspaceEdit
+mkChangeSigEdit uri ss replacement =
+        let txtEdit = TextEdit (realSrcSpanToRange ss) replacement
+            changes = Just $ Map.singleton uri (List [txtEdit])
+        in WorkspaceEdit changes Nothing Nothing
+
+mkNewSignature :: Text -> Text -> Text
+mkNewSignature declName actualType = declName <> " :: " <> actualType
+
+-- | Cleans up type signatures with oddly-named type variables (a0, m0 ...)
```

Comment:
I'm a little unclear why we need to do this. We're using the signature from the message, right? Are they often bad?

Reply:
Yes, Pepe already mentioned there is no reason to tidy the signatures up. I just went overboard. I'm going to drop it.